### DISCUSSION TITLE
what's next in Fore?

### DISCUSSION THREAD

**Discussion by JoernT** [Jinntec/Fore#219] [General]
If you ask yourself what would be next in Fore you may get an impression on our project board which has been updated to contain all topics we are working on.

We are currently working on version 2.0.0 which will introduce some major improvements as well as a ton of fixes, new demos and improved documentation.

There will be two major breaking changes:
* fx-instance will be renamed to fx-data
* fx-submission will be renamed to fx-request

Motivation behind it is that both elements are not reflecting their purpose very well. First question when explaining is always what 'instance' actually means. It is a term that has assiciations to OOP and may lead to wrong assumptions. 

Likewise 'submission' is a term usually used in context with forms and you instantly think of it as something submitting (sending) data while the fx-submission element does much more that just that. It's is more like a descriptive fetch() and fx-request reflects this better. 

There are no plans currently to keep fx-instance and fx-submission beyond 2.0.0 meaning they will instantly been removed. We are aware that this forces users to update their pages but keeping (and just deprecating) the elements would lead to much code-branching and extra effort we'd like to invest into features rather than house-keeping. A simple search/replace will bring you up-to-date with the new version and we thought that this is an acceptable price to pay. 

However if someone objects and has serious arguments we are certainly open to discuss.

Please note that this is a one-time incident - there are no further plans for such renamings of elements.

### Biggest thing to note for Version 2.0.0:

With 'Fore Glass' there will be a brand-new developer tool to inspect the actions, events, the DOM of the page and all associated data. It's completely independent of e.g. Chrome Devtools but can be used in conbination with it. It can be injected into any Fore page by just appending a parameter 'inspect' to the URL.

### Other new features
There will be a couple of new actions and functions as well as many detail improvements which have been batte-tested on several serious applications.

Of course all the details will be listed out in the upcoming release notes. Current plans are to roll it out late summer.

**Comment by timathom** [comment]
Sounds exciting! The name changes make sense to me, but I think you meant "fx-submission will be renamed to fx-request," above (I hope they won't both be called fx-data!)?

Will we see fx-connection in 2.0 as well?

**Reply by JoernT** [reply]
sure

**Reply by JoernT** [reply]
i have fx-connection very basically working but it depends on which i find time within the frame to build a decent demo. So it's still an option but not strictly on the list.

There are also design questions still open.

**Comment by JoernT** [comment]
plan for refactoring has been dropped for now - the effort turned out to be much bigger than expected plus the change just broaden the gap to the original xforms spec.  Although Fore does not seek for full conformance (esp. not using XML but HTML instead) we still keep the orginal element names. Thinking about offering both in a future version

**Reply by JoernT** [reply]
# Update 15. Jan 2024:

## Plans for the year

This is an open list of topics that are planned for 2024 releases:

* support for partial instances (merging in optional elements when necessary for editing)
* performance improvements (still potentials)
* DnD support in and between repeats
* the fx-connection element
* fx-relevant element to switch UI contents on a inscope condition
* many more E2E tests

Please share if you got more ideas for us.

**Reply by timathom** [reply]
+1 to DnD and fx-connection :)

**Reply by JoernT** [reply]
new candidate on the list ...
fx-case with `src` as a shortcut to using the `<fx-load>` action or `<fx-fore src="">`
